File: gtsam-rs/src/nonlinear/isam2.rs
use cxx::UniquePtr;
use sys::ISAM2ResultRust;

use crate::{inference::key::IntoKey, nonlinear::{nonlinear_factor_graph::NonlinearFactorGraph, values::Values}};
use crate::sys::DoubleVec;

pub struct ISAM2 {
    pub(super) inner: UniquePtr<::sys::ISAM2>,
}

impl Default for ISAM2 {
    fn default() -> Self {
        Self {
            inner: ::sys::default_isam2(
                0.1, 10, true, false
            ),
        }
    }
}

impl ISAM2 {
    pub fn new(
        relinearize_threshold: f64, relinearize_skip: i32, cache_linearized_factors: bool, enable_detailed_results: bool
    ) -> Self {
        Self {
            inner: ::sys::default_isam2(
                relinearize_threshold, relinearize_skip, cache_linearized_factors, enable_detailed_results
            ),
        }
    }

    pub fn update(
        &mut self,
        graph: &NonlinearFactorGraph,
        values: &Values,
        new_affected_keys: & Vec<DoubleVec>,
        keys_to_remove: & Vec<u64>,
    ) -> ISAM2ResultRust {
        ::sys::update(self.inner.pin_mut(), &graph.inner, &values.inner, & new_affected_keys, & keys_to_remove)
    }

    pub fn calculate_estimate(
        &self,
    ) -> Values {
        Values {
            inner: ::sys::calculate_estimate(& self.inner),
        }
    }

    pub fn get_marginal_covariance(
        &self,
        key: impl IntoKey,
    ) -> Vec<DoubleVec> {
        ::sys::get_marginal_covariance(& self.inner, key.into_key())
    }
}

File: gtsam-rs/sys/src/nonlinear/isam2.cc
#include "isam2.h"
#include "../base/rust.hpp"

namespace gtsam
{

    std::unique_ptr<ISAM2> default_isam2()
    {
        return std::make_unique<ISAM2>();
    }

    void update_noresults(ISAM2 &isam2, const NonlinearFactorGraph &graph, const Values &initial_values)
    {
        // std::cout << "Initial values: ";
        // initial_values.print();
        // std::cout << std::endl << "Graph: ";
        // graph.print();

        ISAM2Result result = isam2.update(graph, initial_values);
    }

    std::unique_ptr<Values> calculate_estimate(const ISAM2 &isam2) {
        Values estimate = isam2.calculateEstimate();
        std::cout << "SOFIYA! ESTIMATE: ";
        estimate.print();
        return std::make_unique<Values>(estimate);
    }

    rust::Vec<DoubleVec> get_marginal_covariance(
        const ISAM2 &isam2,
        const Key key)
    {
        return eigenmat_to_rustvec(isam2.marginalCovariance(key));
    }
}